internal/app/singularity/oci_linux.go

```diff
@@ -1,87 +1,163 @@
-// Copyright (c) 2018-2020, Sylabs Inc. All rights reserved.
+// Copyright (c) 2018-2022, Sylabs Inc. All rights reserved.
 // This software is licensed under a 3-clause BSD license. Please consult the
 // LICENSE.md file distributed with the sources of this project regarding your
 // rights to use or distribute this software.
 
+// Includes code from https://github.com/containers/podman
+// Released under the Apache License Version 2.0
+
 package singularity
 
 import (
-	"context"
-	"encoding/json"
+	"bufio"
 	"fmt"
+	"io"
 	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/sylabs/singularity/internal/pkg/util/fs"
+	"github.com/sylabs/singularity/internal/pkg/util/user"
+	"github.com/sylabs/singularity/pkg/syfs"
+	"github.com/sylabs/singularity/pkg/util/fs/lock"
+)
 
-	"github.com/sylabs/singularity/internal/pkg/instance"
-	"github.com/sylabs/singularity/internal/pkg/runtime/engine/oci"
-	"github.com/sylabs/singularity/pkg/ociruntime"
-	"github.com/sylabs/singularity/pkg/runtime/engine/config"
-	"github.com/sylabs/singularity/pkg/sylog"
+const (
+	// Absolute path for the runc state
+	RuncStateDir = "/run/singularity-oci"
+	// Relative path inside ~/.singularity for conmon and singularity state
+	ociPath = "oci"
+	// State directory files
+	containerPidFile = "container.pid"
+	containerLogFile = "container.log"
+	runcLogFile      = "runc.log"
+	conmonPidFile    = "conmon.pid"
+	bundleLink       = "bundle"
+	// Files in the OCI bundle root
+	bundleLock   = ".singularity-oci.lock"
+	attachSocket = "attach"
+	// Timeouts
+	createTimeout = 30 * time.Second
 )
 
 // OciArgs contains CLI arguments
 type OciArgs struct {
-	BundlePath     string
-	LogPath        string
-	LogFormat      string
-	SyncSocketPath string
-	PidFile        string
-	FromFile       string
-	KillSignal     string
-	KillTimeout    uint32
-	EmptyProcess   bool
-	ForceKill      bool
+	BundlePath   string
+	LogPath      string
+	LogFormat    string
+	PidFile      string
+	FromFile     string
+	KillSignal   string
+	KillTimeout  uint32
+	EmptyProcess bool
+	ForceKill    bool
+}
+
+// AttachStreams contains streams that will be attached to the container
+type AttachStreams struct {
+	// OutputStream will be attached to container's STDOUT
+	OutputStream io.WriteCloser
+	// ErrorStream will be attached to container's STDERR
+	ErrorStream io.WriteCloser
+	// InputStream will be attached to container's STDIN
+	InputStream *bufio.Reader
```

Comment:
Right. This is some code adapted podman per the comment at the top of this file - I'll double check there's no other use in there (which we might end up borrowing) elsewhere which make the Closer important.

Reply:
Agree this can be simplified as suggested. I think we are unlikely to wrap a HTTP streaming of these, which, from inspection, is the reason for Closers AFAICT.